rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp

```diff
@@ -40,67 +36,36 @@ class MultiThreadedExecutor : public executor::Executor
 public:
   RCLCPP_SMART_PTR_DEFINITIONS(MultiThreadedExecutor);
 
-  MultiThreadedExecutor(memory_strategy::MemoryStrategy::SharedPtr ms =
-    memory_strategies::create_default_strategy())
-  : executor::Executor(ms)
-  {
-    number_of_threads_ = std::thread::hardware_concurrency();
-    if (number_of_threads_ == 0) {
-      number_of_threads_ = 1;
-    }
-  }
+  RCLCPP_PUBLIC
+  MultiThreadedExecutor(
+    memory_strategy::MemoryStrategy::SharedPtr ms = memory_strategies::create_default_strategy());
 
-  virtual ~MultiThreadedExecutor() {}
+  RCLCPP_PUBLIC
+  virtual ~MultiThreadedExecutor();
 
+  RCLCPP_PUBLIC
   void
-  spin()
-  {
-    std::vector<std::thread> threads;
-    {
-      std::lock_guard<std::mutex> wait_lock(wait_mutex_);
-      size_t thread_id_ = 1;  // Use a _ suffix to avoid shadowing `rclcpp::thread_id`
-      for (size_t i = number_of_threads_; i > 0; --i) {
-        std::this_thread::sleep_for(std::chrono::milliseconds(100));
-        auto func = std::bind(&MultiThreadedExecutor::run, this, thread_id_++);
-        threads.emplace_back(func);
-      }
-    }
-    for (auto & thread : threads) {
-      thread.join();
-    }
-  }
+  spin();
 
+  RCLCPP_PUBLIC
   size_t
-  get_number_of_threads()
-  {
-    return number_of_threads_;
-  }
+  get_number_of_threads();
 
-private:
-  void run(size_t this_thread_id)
-  {
-    rclcpp::thread_id = this_thread_id;
-    while (rclcpp::utilities::ok()) {
-      executor::AnyExecutable::SharedPtr any_exec;
-      {
-        std::lock_guard<std::mutex> wait_lock(wait_mutex_);
-        if (!rclcpp::utilities::ok()) {
-          return;
-        }
-        any_exec = get_next_executable();
-      }
-      execute_any_executable(any_exec);
-    }
-  }
+protected:
+  RCLCPP_PUBLIC
+  void
+  run(size_t this_thread_number);
 
+private:
   RCLCPP_DISABLE_COPY(MultiThreadedExecutor);
 
   std::mutex wait_mutex_;
   size_t number_of_threads_;
+  std::unordered_map<std::thread::id, size_t> thread_number_by_thread_id_;
 };
 
-}  // namespace multi_threaded_executor
-}  // namespace executors
-}  // namespace rclcpp
+} /* namespace multi_threaded_executor */
+} /* namespace executors */
+} /* namespace rclcpp */
```

Comment:
I didn't intentionally do that, I was making the alternative transformation everywhere else, but yet:

```
william@farl ~/ros2/src/ros2/rclcpp (git:rclcpp_library:737dba6)
% ament_cpplint /Users/william/ros2/src/ros2/rclcpp/rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp
Using '--root=rclcpp/include' argument

Done processing /Users/william/ros2/src/ros2/rclcpp/rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp

No errors found

william@farl ~/ros2/src/ros2/rclcpp (git:rclcpp_library:737dba6)
% cat /Users/william/ros2/src/ros2/rclcpp/rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp | grep namespace
namespace rclcpp
namespace executors
namespace multi_threaded_executor
} /* namespace multi_threaded_executor */
} /* namespace executors */
} /* namespace rclcpp */
```